File: 5. Stream data/submission/task1.py
import sys
import json
import re
from pyspark import SparkContext
from time import gmtime, strftime
import statistics
from pyspark.streaming import StreamingContext
import binascii
import random
import time

def setone(s):
  ss = int(binascii.hexlify(s.encode('utf8')),16)
  return list(set([(ss*a+b)%c%8000 for a,b,c in hash_tables]))


if __name__ == "__main__":
    start = time.time()
    sc = SparkContext.getOrCreate()
    data = sc.textFile(sys.argv[1])
    city_set = set(data.map(lambda x:json.loads(x)["city"]).distinct().collect())
    nh = 2
    hash_tables = list(zip(random.sample(range(-1000,1000), k=nh),random.sample(range(-10001,10000), k=nh),(5801,10007,300491)))
    #hash_tables = [(693,4838,5801),(518,5404,10007)]
    array_bit = set(data.map(lambda x:json.loads(x)["city"]).distinct().filter(lambda x:x!='').flatMap(setone).distinct().collect())
    ans = []
    with open(sys.argv[2],'r') as file:
        for line in file:
            city = json.loads(line)["city"]
            if city != '':
                ss = int(binascii.hexlify(city.encode('utf8')),16)
                if set([(ss*a+b)%c%8000 for a,b,c in hash_tables]).issubset(array_bit):
                    ans.append("1")
                else:
                    ans.append("0")
            else:
                ans.append("1")
    with open(sys.argv[3],"w") as file:
        file.write(" ".join(ans))
    print("array_bit: ",len(array_bit))
    print(94297-sum(map(int,ans)))
    print("size of ans:",len(ans))
    print(hash_tables)
    end = time.time()
    print("Duration: ",end-start)
            
    

File: 5. Stream data/submission/task1.scala
import org.apache.spark.sql.SparkSession
import java.io._
import scala.math.Ordering.Implicits._
import scala.collection.mutable.HashSet
import scala.collection.mutable.ArrayBuffer
import scala.collection.mutable.ListBuffer
import util.control.Breaks._
import scala.math._
import scala.collection.Map
import org.json4s.jackson.JsonMethods._
import org.json4s._
import org.json4s.jackson.Serialization.write
import org.json4s.JsonDSL._
import java.math.BigInteger

object task1 {
  def setone(s: String,hash_tables:scala.collection.immutable.IndexedSeq[((Int, Int), Int)]): List[Int] = {
      //val ss = s.map(_.toByte).sum
      //val byteString: ByteString = ByteString.copyFrom(DatatypeConverter.parseHexBinary(s))
      val ss = new BigInteger(s.toList.map(_.toInt.toHexString).mkString,16).remainder(BigInteger.valueOf(1000000000)).intValue
      val ans = HashSet.empty[Int]
      for (((a, b), c) <- hash_tables) {
        ans += (((ss * a + b) % c) % 8000).toInt
      }
      ans.toList
    }
  def main(args: Array[String]) {
    val t1 = System.currentTimeMillis
    val ss = SparkSession.builder().appName("scala").config("spark.master", "local[*]").getOrCreate()
    val sc = ss.sparkContext
    sc.setLogLevel("WARN")
    val dd = sc.textFile(args(0)).map(x => (parse(x) \\ "city").values.toString)//sc.textFile("../resource/asnlib/publicdata/business_first.json").map(x => (parse(x) \\ "city").values.toString) //sc.textFile(args(0)).map(x => (parse(x) \\ "city").values.toString)
    val random = new scala.util.Random()
    val nh = 2
    val a = for (i <- 1 to nh) yield -10000+random.nextInt(20000+1)
    val b = for (i <- 1 to nh) yield -10000+random.nextInt(20000+1)
    var c = List(804824051, 24670986, 300491)
    val hash_tables = sc.broadcast(a zip b zip c)
    val city = dd.distinct.collect.toSet

    val array_bit = dd.distinct.filter(x => x != "").flatMap(x=>setone(x,hash_tables.value)).distinct.collect().toSet
    var ans = ArrayBuffer[String]()
    val test = scala.io.Source.fromFile(args(1)).getLines.toList.map(x => (parse(x) \\ "city").values.toString)
    for (ci <- test) {
      if (ci != "") {
        val ss = new BigInteger(ci.toList.map(_.toInt.toHexString).mkString,16).remainder(BigInteger.valueOf(1000000000)).intValue
        //val ss = city.map(_.toByte).sum
        if ((for (((a, b), c) <- hash_tables.value) yield (((ss * a + b) % c) % 8000).toInt).toSet subsetOf array_bit) {
          ans += "1"
          //if (!city.contains(ci)) println("False positive")
        } else {
          ans += "0"
          if (city.contains(ci)) println("False Negative")
        }
      } else {
        ans += "1"
      }
    }
    val file = new File(args(2))
    val output = new BufferedWriter(new FileWriter(file))
    output.write(ans.mkString(" "))
    println(hash_tables.value)
    println("count: %s".format(94297-ans.map(_.toInt).sum))
    println("count of array bit: %s".format(array_bit.size))
    println("size of output: %s".format(ans.size))
    val t2 = System.currentTimeMillis
    println("Duration: %s".format((t2 - t1).toDouble / 1000))
  }

}
